### PR TITLE
Automated diff machine

### PR DESCRIPTION
**Pull request by shawntanzk**
tested it on a test branch and it should work nicely :)

don't know if its the most elegant solution but it's the way i got it to work

### CODE REVIEW

File: src/ontology/cl.Makefile [hunk 1]
```diff
@@ -20,6 +20,11 @@
 #tmp/pr_logical.owl: mirror/pr.owl
 #	echo "Skipped pr logical" && cp $< $@
 
+.PHONY: rdiff
+rdiff: 
+	curl https://raw.githubusercontent.com/shawntanzk/cell-ontology/master/src/ontology/cl-edit.owl -o rdifftmp.owl
```
Comment (matentzn) [inline src/ontology/cl.Makefile:25]: best do not curl (unnecessary dependency on network - always avoid), use `checkout master -- $(SRC)` instead.

File: src/ontology/cl.Makefile [hunk 2]
```diff
@@ -20,6 +20,11 @@
...
+	$(ROBOT) diff -labels True -left rdifftmp.owl --right cl-edit.owl -format markdown --output diff.md
```
Comment (matentzn) [inline src/ontology/cl.Makefile:26]: For now this is fine @ubyndr it would be important that however you realised the final diff system we integrate in ODK, that it can be run on _any_ file in the ODK pipeline.

**Review by matentzn — CHANGES_REQUESTED** [review]

File: src/ontology/cl.Makefile (hunk 1 above)
Comment (shawntanzk) [reply src/ontology/cl.Makefile:25]: thanks! this was the part that I was quite unhappy about how inelegant it was, i usually use git show for this part but didnt know the best way to do this otherwise :)

Comment (matentzn) [reply src/ontology/cl.Makefile:25]: git show may be fine, maybe even better, I have not used it - just do not curl!

File: src/ontology/cl.Makefile (hunk 2 above)
Comment (dosumis) [reply src/ontology/cl.Makefile:26]: Can we make this dependent on file extension?

File: src/ontology/cl.Makefile (hunk 1 above)
Comment (shawntanzk) [reply src/ontology/cl.Makefile:25]: the struggle I had with that was that I couldn't get git commands to work with actions - trying this on test fork and it seems to be failing there too. but 100% agree that curl is not ideal

File: src/ontology/cl.Makefile (hunk 2 above)
Comment (matentzn) [reply src/ontology/cl.Makefile:26]: exactly.. I agree. It should be something like:

```
rdiff-%: tmp/main-% tmp/branch-%*
	ROBOT diff tmp/main-$* tmp/branch-$* ....
```

Comment (matentzn) [reply src/ontology/cl.Makefile:26]: and if it is, say cl.owl, it should depend on cl.owl so that it is actually created by make first!

File: src/ontology/cl.Makefile (hunk 1 above)
Comment (matentzn) [reply src/ontology/cl.Makefile:25]: I think the cleanest thing is to orchestrate this in the action itself, have multiple steps with different checkouts, each generating the file in question, then running the diff in a third step.

Comment (balhoff) [reply src/ontology/cl.Makefile:25]: @shawntanzk here's what I implemented for GO a while back, maybe it can be of use: https://github.com/geneontology/go-ontology/blob/master/.github/workflows/inference_diff.yml

Comment (matentzn) [reply src/ontology/cl.Makefile:25]: Oh YES, that is right you did that.. Sorry man I forgot about it. I think this is the best design!

Comment (shawntanzk) [reply src/ontology/cl.Makefile:25]: sweet! thanks so much @balhoff :D

File: src/ontology/cl.Makefile (hunk 2 above)
Comment (shawntanzk) [reply src/ontology/cl.Makefile:26]: does this mean we want to compare $(SRC).owl that is made from $(SRC)-edit.owl?

File: src/ontology/cl.Makefile (hunk 1 above)
Comment (ubyndr) [reply src/ontology/cl.Makefile:25]: Do we need 4 jobs, classify_pr, classify_master, diff_classification and post_comment? We are checking out our branch 3 times is that necessary? Can I use single job for, downloading branch, master, doing diff analysis and posting a comment? Or is that a bad design? @matentzn

**Comment by shawntanzk** [comment]
@matentzn & @dosumis - current way from @balhoff creates a comment every commit - is this ideal or should we use a method that updates the first comment each time?

**Comment by matentzn** [comment]
Creating a comment at every committ seems correct to me;

**Comment by dosumis** [comment]
I think I'd prefer to update a single comment otherwise thread may get very long.  It has the downside that comments might make a bit less sense as updates happen.  But I think that's OK.

**Comment by shawntanzk** [comment]
> I'd prefer to update a single comment 
personally this is my preference too, for that reason

**Comment by ubyndr** [comment]
I have change the robot comment that @balhoff used in his action 
`robot diff --left ../../reasoned-master.owl/reasoned.owl --left-catalog catalog-v001.xml --right ../../reasoned-pr.owl/reasoned.owl --right-catalog catalog-v001.xml -f markdown -o classification-diff.md`, 
to 
`robot diff --left master/src/ontology/cl-edit.owl --right src/ontology/cl-edit.owl -f markdown -o classification-diff.md`. 
What is the purpose of using left and right catalog? Is that necessary? @matentzn & @dosumis

**Comment by matentzn** [comment]
Ah sorry i didn't read that right. Single comment for sure.. I thought the question was how often to update, forget what I said above.

**Comment by shawntanzk** [comment]
> What is the purpose of using left and right catalog?

This was my question when I saw it too lol

**Comment by shawntanzk** [comment]
seems --update=TRUE either 1) doesnt work, or 2) is super confused cause there are so many previous comments - @ubyndr to test this out on another ontology in a new PR

**Comment by balhoff** [comment]
If you don't provide the catalog, the imports won't be loaded correctly from local files.

File: src/ontology/cl.Makefile (hunk 1 above)
Comment (balhoff) [reply src/ontology/cl.Makefile]: The separate jobs made it easier (to me) to work on two different branches. It could be 3 total jobs, but the diff and the comment used different docker images, so they are different jobs. But it sounds like you found a better comment posting action. Also with multiple jobs, the two classifications can happen simultaneously.

**Comment by shawntanzk** [comment]
> If you don't provide the catalog, the imports won't be loaded correctly from local files.

@ubyndr - yeah we should put the catalog file then :) thanks @balhoff

**Comment by shawntanzk** [comment]
Strategy: 
Change the ROBOT diff part to makefile and use github actions to call command in makefile
use variable for what files that the actions will look into (do this in ONT.makefile)

Strech: for ODK make a config that can define the files to go through diff

Thoughts - can the current commenting thing have 3 different comment? -> check if this is possible

**Comment by shawntanzk** [comment]
minimum check these 2: 
ONT-edit.editformat 
ONT.owl

File: .github/workflows/diff.yml [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+      - uses: actions/checkout@v2
+      - name: Classify ontology
+        run: cd src/ontology; make IMP=FALSE PAT=FALSE MIR=FALSE cl.owl
+      - name: Upload PR cl.owl
+        uses: actions/upload-artifact@v2
+        with:
+          name: cl-pr.owl
+          path: src/ontology/cl.owl
+          retention-days: 1
+  classify_master:
```
Comment (matentzn) [inline .github/workflows/diff.yml:43]: We avoid `master` nomenclature now for political correctness reasons, use `classify_main` and `classify_branch` instead.

File: .github/workflows/diff.yml [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        run: export ROBOT_JAVA_ARGS=-Xmx6G; cd src/ontology; robot diff --left cl-master.owl/cl.owl --left-catalog catalog-v001.xml --right cl-pr.owl/cl.owl --right-catalog catalog-v001.xml -f markdown -o classification-diff.md
+      - name: Upload diff
+        uses: actions/upload-artifact@v2
+        with:
+          name: classification-diff.md
+          path: src/ontology/classification-diff.md
+  post_comment:
+    needs: [diff_classification, edit_file]
+    runs-on: ubuntu-latest
+    container: docker://dvcorg/cml-py3:latest
```
Comment (matentzn) [inline .github/workflows/diff.yml:86]: Why is this needed? Out of curiosity.

**Review by matentzn — APPROVED** [review]
Awesome.. Looks impressive! I would change the nomenclature throughout (we don't use the term `master` anymore unless we have to, for example in the branches: section).

File: .github/workflows/diff.yml (hunk 4 above)
Comment (shawntanzk) [reply .github/workflows/diff.yml:86]: oh whoops this isn't, was testing with something and totally forgot to remove it. @ubyndr removed in here the last time, but I forgot to remove it on my test fork and copied and paste the whole thing - will remove

**Comment by shawntanzk** [comment]
> Do we really need to use obolibrary/odkfull:v1.2.32

pretty sure we need it for the make step, the rest I think lite is fine

**Comment by matentzn** [comment]
odklite is sufficient

**Comment by shawntanzk** [comment]
If all is good, should I merged this in for CL or just wait for roll out with ODK?

**Comment by matentzn** [comment]
Either way is fine. I will compile it soon and make a PR on CL

**Comment by shawntanzk** [comment]
Will merge this in then :) no harm and all
